factcast-client-grpc/src/main/java/org/factcast/client/grpc/ResilientGrpcSubscription.java

```diff
@@ -0,0 +1,298 @@
+    }
+  }
+
+  @VisibleForTesting
+  ResilientGrpcSubscription delegate(Consumer<Subscription> consumer) {
+    for (; ; ) {
+      assertSubscriptionStateNotClosed();
+      try {
+        Subscription cur = currentSubscription.getAndBlock();
+        consumer.accept(cur);
+        return this;
+      } catch (Exception e) {
+        if (!resilience.shouldRetry(e)) {
+          throw ExceptionHelper.toRuntime(e);
+        }
+      }
+    }
+  }
+
+  private void assertSubscriptionStateNotClosed() {
+    if (isClosed.get()) {
+      throw new SubscriptionClosedException(
+          "Subscription already closed  (" + originalRequest + ")");
+    }
+  }
+
+  private synchronized void connect() {
+    log.debug("Connecting ({})", originalRequest);
+    reConnect();
+  }
+
+  @VisibleForTesting
+  synchronized void reConnect() {
+    log.debug("Reconnecting ({})", originalRequest);
+    SubscriptionRequestTO to = SubscriptionRequestTO.forFacts(originalRequest);
+    UUID last = lastFactIdSeen.get();
+    if (last != null) {
+      to.startingAfter(last);
+    }
+
+    if (currentSubscription.get() == null) {
+      try {
+        Subscription plainSubscription = store.internalSubscribe(to, delegatingObserver);
+        currentSubscription.set(plainSubscription);
+      } catch (Exception e) {
+        fail(e);
+      }
+    }
+  }
+
+  private void closeAndDetachSubscription() {
+    Subscription current = currentSubscription.getAndSet(null);
+    try {
+      if (current != null) current.close();
+    } catch (Exception justLog) {
+      log.warn("Ignoring Exception while closing a subscription ({})", originalRequest, justLog);
+    }
+  }
+
+  @VisibleForTesting
+  void fail(Throwable exception) {
+    log.error("Too many failures, giving up. ({})", originalRequest);
+    close();
+    currentSubscription.unblock();
+    originalObserver.onError(exception);
+    throw ExceptionHelper.toRuntime(exception);
+  }
+
+  @FunctionalInterface
+  interface ThrowingBiConsumer<T, U> {
+    void accept(T t, U u) throws TimeoutException;
+  }
+
+  class DelegatingFactObserver implements FactObserver {
+    @Override
+    public void onNext(@NonNull Fact element) {
+      if (!isClosed.get()) {
+        originalObserver.onNext(element);
+        lastFactIdSeen.set(element.id());
+      } else {
+        log.warn("Fact arrived after call to .close() [a few of them is ok...]");
+      }
+    }
+
+    @Override
+    public void onCatchup() {
+      originalObserver.onCatchup();
+    }
+
+    @Override
+    public void onComplete() {
+      originalObserver.onComplete();
+    }
+
+    @Override
+    public void onFastForward(@NonNull UUID factIdToFfwdTo) {
+      originalObserver.onFastForward(factIdToFfwdTo);
+    }
+
+    @Override
+    public void onError(@NonNull Throwable exception) {
+      log.info("Closing subscription due to onError triggered.  ({})", originalRequest, exception);
+      closeAndDetachSubscription();
+
+      resilience.registerAttempt();
```

Comment:
Shouldnt this be inside the if? 🤔

Reply:
depends on how you interpret resilience.retries = 10
is this going to be fail after 10 failures or after 11?

Thinking about it, should we rename the property to attempts ?